### PR TITLE
4616: Order table foreign keys

### PR DESCRIPTION
**Pull request by RuthBerry96**
Story: [AB#4616](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/4616)
Task: [AB#7170](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/7170)

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql [hunk 1]
```diff
@@ -15,6 +15,7 @@
     LastUpdatedByName NVARCHAR(256) NULL,     
 
     CONSTRAINT FK_Order_OrderStatusId_OrderStatus_OrderStatusId FOREIGN KEY (OrderStatusId) REFERENCES OrderStatus (OrderStatusId), 
-    CONSTRAINT FK_Order_AddressId_Address_AddressId FOREIGN KEY (OrganisationAddressId) REFERENCES Address (AddressId),
+    CONSTRAINT FK_Order_OrganisationAddressId_Address_AddressId FOREIGN KEY (OrganisationAddressId) REFERENCES Address (AddressId),
+    CONSTRAINT FK_Order_OrganisationBillingAddressId_Address_AddressId FOREIGN KEY (OrganisationBillingAddressId) REFERENCES Address (AddressId),
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql:19]: Maybe slap some square brackets around Address.

It's not actually a SQL keyword but VS likes to highlight it for some reason...

File: src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql [hunk 2]
```diff
@@ -15,6 +15,7 @@
     LastUpdatedByName NVARCHAR(256) NULL,     
 
     CONSTRAINT FK_Order_OrderStatusId_OrderStatus_OrderStatusId FOREIGN KEY (OrderStatusId) REFERENCES OrderStatus (OrderStatusId), 
-    CONSTRAINT FK_Order_AddressId_Address_AddressId FOREIGN KEY (OrganisationAddressId) REFERENCES Address (AddressId),
+    CONSTRAINT FK_Order_OrganisationAddressId_Address_AddressId FOREIGN KEY (OrganisationAddressId) REFERENCES Address (AddressId),
+    CONSTRAINT FK_Order_OrganisationBillingAddressId_Address_AddressId FOREIGN KEY (OrganisationBillingAddressId) REFERENCES Address (AddressId),
     CONSTRAINT FK_Order_OrganisationContactId_Contact_ContactId FOREIGN KEY (OrganisationContactId) REFERENCES Contact (ContactId)
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql:20]: I think the FK names are a tad long and slightly confusing although I'm struggling to offer any alternatives. My preference would be to try to simplify them somehow. Perhaps something like below?

```sql
FK_Order_OrderStatus
FK_Order_OrganisationAddress
FK_Order_OrganisationBillingAddress
FK_Order_OrganisationContact
```

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs [hunk 3]
```diff
@@ -1,76 +1,76 @@
...
-        {
-            if (order is null)
-            {
-                throw new ArgumentNullException(nameof(order));
-            }
-
-            using ( var dbContextTransaction = _context.Database.BeginTransaction())
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:3]: Do we need this?

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs [hunk 4]
```diff
@@ -1,76 +1,76 @@
...
+        }
+
+        public async Task<string> CreateOrderAsync(Order order)
+        {
+            if (order is null)
+            {
+                throw new ArgumentNullException(nameof(order));
+            }
+
+            using (var dbContextTransaction = _context.Database.BeginTransaction())
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:59]: Can we use `BeginTransactionAsync` here?

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs [hunk 5]
```diff
@@ -1,76 +1,76 @@
...
-
-            return order.OrderId;
+                await _context.SaveChangesAsync();
+                await dbContextTransaction.CommitAsync();
+            }
+
+            return order.OrderId;
         }
 
         private async Task<string> GetIncremenedOrderId()
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:76]: While you're here would you mind tidying up this method?

The method name should be `GetIncrementedOrderId`.

Could you also tidy up the call to `int.Parse` by removing the redundant argument name (`s:`) and simplify `System.Globalization.CultureInfo.InvariantCulture` by introducing a `using`?

I'd also be tempted to remove the `using static System.Int32;` and include the type name instead, e.g. `int.Parse`.

File: src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql (hunk 2 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql:20]: 👍

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs (hunk 3 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:3]: nope will remove

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs (hunk 5 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:76]: 👍

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs (hunk 4 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:59]: 👍

File: src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql (hunk 1 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql:19]: 👍

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs [hunk 6]
```diff
@@ -56,31 +55,31 @@ public async Task<string> CreateOrderAsync(Order order)
                 throw new ArgumentNullException(nameof(order));
             }
 
-            using (var dbContextTransaction = _context.Database.BeginTransaction())
+            using (var dbContextTransaction = _context.Database.BeginTransactionAsync())
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:58]: I think we probably want an `await` here.

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs [hunk 7]
```diff
@@ -56,31 +55,31 @@ public async Task<string> CreateOrderAsync(Order order)
...
-                    order.OrderId = await GetIncremenedOrderId();
+                    order.OrderId = await GetIncrementedOrderId();
                 }
 
                 order.OrderStatus = await _context.OrderStatus.FindAsync(order.OrderStatus.OrderStatusId) ?? order.OrderStatus;
 
                 _context.Order.Add(order);
                 await _context.SaveChangesAsync();
-                await dbContextTransaction.CommitAsync();
+                await (await dbContextTransaction).CommitAsync();
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:69]: I think you'll be able to remove this `await` if you add one for `BeginTransactionAsync`.

File: src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql [hunk 8]
```diff
@@ -16,9 +16,12 @@
...
-    CONSTRAINT FK_Order_OrganisationContactId_Contact_ContactId FOREIGN KEY (OrganisationContactId) REFERENCES Contact (ContactId)
+    LastUpdatedByName NVARCHAR(256) NULL,   
+    
+    CONSTRAINT FK_Order_OrderStatus FOREIGN KEY (OrderStatusId) REFERENCES OrderStatus (OrderStatusId),    
+    CONSTRAINT FK_Order_OrganisationAddress FOREIGN KEY (OrganisationAddressId) REFERENCES [Address] (AddressId),
+    CONSTRAINT FK_Order_OrganisationBillingAddress FOREIGN KEY (OrganisationBillingAddressId) REFERENCES [Address] (AddressId),     
+    CONSTRAINT FK_Order_OrganisationContact FOREIGN KEY (OrganisationContactId) REFERENCES Contact (ContactId), 
+    CONSTRAINT FK_Order_SupplierAddress FOREIGN KEY (SupplierAddressId) REFERENCES [Address] (AddressId),     
+    CONSTRAINT FK_Order_SupplierContact FOREIGN KEY (SupplierContactId) REFERENCES Contact (ContactId)
 );
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql:27]: This isn't essential (and is just me being exceptionally picky!) but I think there's a bit of trailing white space on a few lines here.

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs (hunk 6 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:58]: 👍

File: src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs (hunk 7 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.Persistence/Repositories/OrderRepository.cs:69]: 👍

File: src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql (hunk 8 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Ordering.OrderingDatabase/Tables/Order.sql:27]: 👍

**Pull request merged**